x-pack/plugins/fleet/server/services/agents/action_runner.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import uuid from 'uuid';
+import type { SortResults } from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient, SavedObjectsClientContract } from '@kbn/core/server';
+import { withSpan } from '@kbn/apm-utils';
+
+import { isResponseError } from '@kbn/es-errors';
+
+import type { Agent, BulkActionResult } from '../../types';
+import { appContextService } from '..';
+import { SO_SEARCH_LIMIT } from '../../../common/constants';
+
+import { getAgentActions } from './actions';
+import { closePointInTime, getAgentsByKuery } from './crud';
+
+export interface ActionParams {
+  kuery: string;
+  showInactive?: boolean;
+  batchSize?: number;
+  total?: number;
+  actionId?: string;
+  // additional parameters specific to an action e.g. reassign to new policy id
+  [key: string]: any;
+}
+
+export interface RetryParams {
+  pitId: string;
+  searchAfter?: SortResults;
+  retryCount?: number;
+  taskId?: string;
+}
+
+export abstract class ActionRunner {
+  protected esClient: ElasticsearchClient;
+  protected soClient: SavedObjectsClientContract;
+
+  protected actionParams: ActionParams;
+  protected retryParams: RetryParams;
+
+  constructor(
+    esClient: ElasticsearchClient,
+    soClient: SavedObjectsClientContract,
+    actionParams: ActionParams,
+    retryParams: RetryParams
+  ) {
+    this.esClient = esClient;
+    this.soClient = soClient;
+    this.actionParams = { ...actionParams, actionId: actionParams.actionId ?? uuid() };
+    this.retryParams = retryParams;
+  }
+
+  protected abstract getActionType(): string;
+
+  protected abstract getTaskType(): string;
+
+  protected abstract processAgents(agents: Agent[]): Promise<{ items: BulkActionResult[] }>;
+
+  /**
+   * Common runner logic accross all agent bulk actions
+   * Starts action execution immeditalely, asynchronously
+   * On errors, starts a task with Task Manager to retry max 3 times
+   * If the last batch was stored in state, retry continues from there (searchAfter)
```

Comment:
I wonder if we should be checkpointing the state regardless of whether or not there is an error. We need to consider the case where Kibana crashes or is otherwise shutdown. Ideally in this case there's always a task scheduled to confirm that the process was completed successfully. This may require some significant changes to this PR and if so, we could handle it in a follow up.

Reply:
yes, I'll tackle this separately.